### DISCUSSION TITLE
Refresh Access Token expiry after 14 days

### DISCUSSION THREAD

**Discussion by rangan-anand** [AzureAD/microsoft-authentication-library-for-python#386] [Q&A]
Hello,

I have a daemon app in python which runs in AWS lambda this also have subscription enabled on Inbox(whenever a new mail comes in the Mailbox this app will process the data and load onto a table in backend), and it connects to token cache to access the refresh token to access Graph API, all the setup works without any issue, but after 14 days of initial login, the refresh access token expires automatically and it needs user intervention to authenticate via initiate_device_flow.

I am looking for an alternate approach which reduces the User login. please let me know if there is  a workaround?

**Comment by rayluo** [comment]
I found a Q&A here ["For how long I can keep using the refresh token?"](https://docs.microsoft.com/en-us/answers/questions/340205/for-how-long-i-can-keep-using-the-refresh-token.html) (which provides even more detailed information than our [standard document](https://docs.microsoft.com/en-us/azure/active-directory/develop/refresh-tokens#refresh-token-lifetime)).

Could it be possible that your app implementation manage your refresh token (RT) but did not persist subsequent new RT? Does your app utilize MSAL's built-in token cache behavior?

**Reply by rangan-anand** [reply]
Hi,
I ran the app today and its using the built-in cache behaviour where the "credential_type": "RefreshToken" is getting updated along with the "last_modification_time" set to today.  

could it be possible that this is because of other conditional policy settings on an organization level, which would require the users to login every 14 days? please let me know.

attaching the detailed error below.

{'error': 'invalid_grant', 'error_description': 'AADSTS70043: The refresh token has expired or is invalid due to sign-in frequency checks by conditional access. The token was issued on 2021-07-09T15:15:21.4921392Z and the maximum allowed lifetime for this request is 1209600.\r\nTrace ID: ''\r\nCorrelation ID: ''\r\nTimestamp: 2021-07-24 12:22:13Z', 'error_codes': [70043], 'timestamp': '2021-07-29 12:22:13Z', 'trace_id': '', 'correlation_id': '', 'suberror': 'token_expired', 'classification': ''}

**Reply by rayluo** [reply]
What that error message said

> "AADSTS70043: The refresh token has expired or is invalid due to sign-in frequency checks by conditional access. The token *was issued on 2021-07-09T15:15:21.4921392Z* and the maximum allowed lifetime for this request is 1209600."

matches the Q&A link above which suggests the sliding window is good for 14 days.

Has your app *not* been run since 2021-7-9? Or is the token cache *not* actually persisted onto disk?

**Reply by rangan-anand** [reply]
Thanks for the quick response, this error came in on 2021-7-9, after that we have been updating the token cache every 13 days to avoid failure,
I have checked the token cache file and updated RT is getting persisted onto the disk,

**Reply by rayluo** [reply]
> we have been updating the token cache every 13 days to avoid failure,

Sounds like you ended up using that workaround and then your app works? You were only asking the "14 days expiration" topic here for clarification?

Regarding the implementation, MSAL's `acquire_token_silent()` would already update the refresh token (RT). However, the default token cache is in-memory, therefore MSAL won't actually persist the new RT. This behavior and its [solution is documented in the sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.13.0/sample/confidential_client_secret_sample.py#L48-L50). If you use such a recommended pattern, you shouldn't need to know those "14 days sliding window" details and write your app for that.

**Reply by rangan-anand** [reply]
yes, we are using a workaround to avoid the failure for now, I am trying to understand the issue on the app setup so that the workaround could be avoided as we are trying to scale up the app to support more users.

setup of the app is as below 

```
persistence = FilePersistence('token_cache.bin')
app = PublicClientApplication(client_id=client_id,
                         authority=authority,
                         token_cache=PersistedTokenCache(persistence))
```

instead of using msal.SerializableTokenCache , I am using msal_extensions to achieve the similar result.

also I am using MSAL's acquire_token_silent_with_error() to get access token, which would update the (RT) in cache which is maintained in a file, however the app fails every 14 days saying the below error, could you please let me know if i have to make additional changes to avoid this failure?

The refresh token has expired or is invalid due to sign-in frequency checks by conditional access. The token was issued on 'some date' and the maximum allowed lifetime for this request is 1209600

also i noticed that on the token cache file, RT is getting updated frequently but not the Access token , is this the expected behaviour?

**Reply by rayluo** [reply]
> also i noticed that on the token cache file, RT is getting updated frequently but not the Access token , is this the expected behaviour?

This one is particularly weird. While it does not exactly match the expected the behaviour (more on this later), an updated RT should at least avoid your hitting that 14-day slide window mechanism.

In such a case, some investigation may be needed. You could use the aforementioned [sample app](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.13.0/sample/confidential_client_secret_sample.py#L48-L50) (it is OK to swap its `acquire_token_silent()` with `acquire_token_silent_with_error()`), together with `SerializableTokenCache` or `FilePersistence` (they should be interchangeable here, but it doesn't harm to test both of them, one after another, to rule out or pinpoint some issue). Then you try acquire token for your same scope.

The expected behavior is:

1. Token cache is initially empty.
2. Token cache would contain access token (AT) and refresh token (RT) after first successful `acquire_token_by_xyz()` call. (BTW, which `acquire_token_by_xyz()` does your app use?)
3. Call `acquire_token_silent()` (or `acquire_token_silent_with_error()`) with same scope within an hour. AT in cache should be returned, and the token cache file should keep same AT and RT.
4. After one hour, the AT would typically expire. Re-do step 3, a new AT should be obtained and cached. The token cache should be updated with that new AT, AND *typically a new RT, too*.

**Reply by rangan-anand** [reply]
Thanks, is it possible that there is some kind of conditional access policy set on tenant level which triggers this expiry every 14 days,

Thanks in advance.

**Reply by rayluo** [reply]
You asked the [same question 15 days ago](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/386#discussioncomment-1119211). At that time, our follow-up conversation went like this.

> > > we have been updating the token cache every 13 days to avoid failure,
> > 
> > Sounds like you ended up using that workaround and then your app works? You were only asking the "14 days expiration" topic here for clarification?
> >
> > Regarding the implementation, MSAL's `acquire_token_silent()` would already update the refresh token (RT). However, the default token cache is in-memory, therefore MSAL won't actually persist the new RT. This behavior and its [solution is documented in the sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.13.0/sample/confidential_client_secret_sample.py#L48-L50). If you use such a recommended pattern, you shouldn't need to know those "14 days sliding window" details and write your app for that.
>
> yes, we are using a workaround to avoid the failure for now, I am trying to understand the issue on the app setup so that the workaround could be avoided

My understanding is, the fact that your workaround of updating the tokens every 13 days can bypass the 14-day refresh token (RT) valid window, seems to suggest that your app without your workaround somehow did not update the RT. Can you first double check that aspect?

**Reply by rangan-anand** [reply]
Hi Rayluo,
the acquire_token_silent_with_error wasn't updating the Access token in token cache persisted on file , so added force_refresh=True now its updating the token cache with the latest access token along with refresh token(RT), 

would conditional access have any impact on the token expiry after 14 days, or should this work without any hitch now that i have persisted the RT and access token? please let me know your thoughts. Thanks in advance.

**Reply by rayluo** [reply]
> Hi Rayluo,
> the acquire_token_silent_with_error wasn't updating the Access token in token cache persisted on file , so added force_refresh=True now its updating the token cache with the latest access token along with refresh token(RT),

In [this message](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/386#discussioncomment-1126305) we went through the expected behavior of MSAL's token cache. If your observation "the acquire_token_silent_with_error wasn't updating the Access token in token cache persisted on file" was happening at step No.3 i.e. during the first hour of a newly acquired access token (AT), then it is a correct and desirable behavior.

Always adding `force_refresh=True` defeats the purpose of a token cache. (That flag only exists for historical and backward compatibility reason. In all other normal cases, you shouldn't need to do `force_refresh=True`.)

> 
> would conditional access have any impact on the token expiry after 14 days, or should this work without any hitch now that i have persisted the RT and access token? please let me know your thoughts. Thanks in advance.

The confusing part to me is you said "*now that* I have persisted ...". Didn't your earlier message indicate that you've long been using `FilePersistence` already? Such a discrepancy in observation would normally indicate bugs in the app's implementation.

In the [same message](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/386#discussioncomment-1126305) above, it also referred to a sample. Together with `SerializableTokenCache` or `FilePersistence` it would have already persisted RT and AT by default. Would you mind try running that sample hourly to observe it indeed updates the RT and AT for you, so that we would be confident that the sample would not have the 14-day issue in the first place?

**Reply by rayluo** [reply]
This troubleshooting conversation became inactive. But to recap our understanding:
* AT will expire soon, RT will also eventually expire, but calling MSAL's `acquire_token_silent()` will give you a fresh-enough AT whenever possible, and it will automatically maintain the fresh AT & RT in the token cache.
* You should not need to force_refresh.
